eventlog2speedscope/src/Speedscope/Main.hs

```diff
@@ -367,30 +396,35 @@ emptyThreadAnalysis =
 
 analyzeThread
   :: MonadSafe m
-  => Connection
+  => FrameFilter
+  -> Connection
   -> GHC.ThreadId
   -> Producer FrameEvent m ()
-analyzeThread conn threadId =
+analyzeThread frameFilter conn threadId =
   do
     ref <- newIORef emptyThreadAnalysis & Pipes.liftIO
     Pipes.for (produceThreadEvents conn threadId)
-      (\event -> processThreadEvent event & Pipes.hoist (refStateT ref))
+      (\event -> processThreadEvent frameFilter event & Pipes.hoist (refStateT ref))
     pure ()
 
 processThreadEvent
   :: MonadState ThreadAnalysis m
-  => ThreadEvent
+  => FrameFilter
+  -> ThreadEvent
   -> Producer FrameEvent m ()
-processThreadEvent ThreadEvent { frameEvent } = threadFrame frameEvent
+processThreadEvent frameFilter ThreadEvent { frameEvent } =
+    threadFrame frameFilter frameEvent
 
 threadFrame
   :: MonadState ThreadAnalysis m
-  => FrameEvent
+  => FrameFilter
+  -> FrameEvent
   -> Producer FrameEvent m ()
-threadFrame frameEvent@FrameEvent { eventType, eventFrameId } =
+threadFrame frameFilter frameEvent@FrameEvent { eventType, eventFrameId } =
   do
     case eventType of { Open -> pushThread; Close -> popThread } eventFrameId
-    Pipes.yield frameEvent
+    threadAnalysis <- State.get
+    when (applyFrameFilter frameFilter threadAnalysis eventFrameId) (Pipes.yield frameEvent)
```

Comment:
By this point, the stack in `ThreadAnalysis` has been updated: an `Open` event pushed a new frame onto the stack or a `Close` event popped one off. Because of that, in `applyFrameFilter`, it's important that we only look at the _tail_ of the stack and ignore the top-most entry. Otherwise, `--not-matching-children` will skip the `Open` event of the parent but it will still try to emit the `Close` event. This is the point where speedscope complains.